alfresco-java-rest-api-common/src/main/java/org/alfresco/rest/sdk/feign/oauth2/OAuth2FeignRequestInterceptor.java

```diff
@@ -17,45 +17,187 @@
     private final OAuth2AuthorizeRequest oAuth2AuthorizeRequest;
+    private final OAuth2AuthorizedClientService authorizedClientService;
 
     /**
      * Constructor.
      *
      * @param oAuth2AuthorizedClientManager given {@link OAuth2AuthorizedClientManager}
      * @param oAuth2AuthorizeRequest        given {@link OAuth2AuthorizeRequest}
+     * @param authorizedClientService       given {@link OAuth2AuthorizedClientService}
      */
-    public OAuth2FeignRequestInterceptor(OAuth2AuthorizedClientManager oAuth2AuthorizedClientManager, OAuth2AuthorizeRequest oAuth2AuthorizeRequest) {
+    public OAuth2FeignRequestInterceptor(OAuth2AuthorizedClientManager oAuth2AuthorizedClientManager,
+                                         OAuth2AuthorizeRequest oAuth2AuthorizeRequest,
+                                         OAuth2AuthorizedClientService authorizedClientService)
+    {
         this.oAuth2AuthorizedClientManager = oAuth2AuthorizedClientManager;
         this.oAuth2AuthorizeRequest = oAuth2AuthorizeRequest;
+        this.authorizedClientService = authorizedClientService;
     }
 
     @Override
-    public void apply(RequestTemplate template) {
+    public void apply(RequestTemplate template)
+    {
         LOGGER.debug("Applying OAuth2 authentication to feign request template {}", template);
         template.header(AUTHORIZATION, getAuthorizationToken());
     }
 
-    private String getAuthorizationToken() {
-        final OAuth2AccessToken accessToken = oAuth2AuthorizedClientManager.authorize(oAuth2AuthorizeRequest).getAccessToken();
-        return String.format(AUTH_HEADER_FORMAT, accessToken.getTokenType().getValue(), accessToken.getTokenValue());
+    /**
+     * Retrieves the authorization token. Checks if the access token is valid and refreshes it if expired.
+     * If no token exists, it will trigger the first-time authorization flow.
+     *
+     * @return the formatted authorization token string
+     */
+    private String getAuthorizationToken()
+    {
+        OAuth2AuthorizedClient client = getAuthorizedClient();
+        OAuth2AccessToken accessToken;
+        if (client != null && client.getAccessToken() != null)
+        {
+            accessToken = client.getAccessToken();
+
+            // Check if the access token has expired
+            if (isAccessTokenExpired(accessToken))
+            {
+                LOGGER.info("Access token expired. Attempting to refresh...");
+                // Try to refresh the access token using the refresh token
+                client = refreshAccessToken(client);
+                accessToken = client.getAccessToken();
+            }
+        }
+        else
+        {
+            // If no token is found, we need to initiate the first-time authorization flow
+            LOGGER.info("Authorized Client not found. Initiating OAuth2 authorization flow...");
+            client = startAuthorizationFlow();
```

Comment:
Are we always going to follow the Authorization Code Flow or Alfresco Java SDK follows other Oauth2 Grant types as well?

Reply:
Alfresco-java-sdk follows grant-type ='password' and 'referesh-token'. Incase of failure using grant-type 'referesh-token' it will use grant-type password again to get access and referesh token.